### PR TITLE
[bugfix](stdcallonce) replace std callonce with a lock because it is not exception safe

### PR DESCRIPTION
**Pull request by yiguolei**
## Proposed changes

According to cppreference https://en.cppreference.com/w/cpp/thread/call_once. If an exception throws from the call method, the flag is not set and the caller method could catch the exception and handle it.

But I tried with clang16 and gcc11。 And also sometime clang17 will core. The core stack is as follows:

img_v3_02b3_2eb61bac-8dcd-4de6-aad9-51e6d188f5bg

So that doris should not depend on it.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yiguolei** [comment]
run buildall

File: be/src/util/once.h [hunk 1]
```diff
@@ -77,6 +99,7 @@ class DorisCallOnce {
 private:
     std::atomic<bool> _has_called;
     std::once_flag _once_flag;
```
Comment (gavinchou) [inline be/src/util/once.h:101]: Is _once_flag still needed?

Comment (yiguolei) [reply be/src/util/once.h]: Useless, I removed it.

**Comment by yiguolei** [comment]
run buildall

File: be/src/util/once.h [hunk 2]
```diff
@@ -52,14 +52,36 @@ class DorisCallOnce {
...
+            return _status;
+        }
+        std::lock_guard l(_flag_lock);
+        // After lock, should check again. Another thread may call
+        // at the same time during first time check has called.
+        if (has_called()) {
+            return _status;
+        }
+        _status = fn();
+        _has_called.store(true, std::memory_order_release);
```
Comment (zclllyybb) [inline be/src/util/once.h:84]: should `store` then `fn()`?

**Review by zclllyybb — CHANGES_REQUESTED** [review]
In test should test concurrency thoroughly. maybe just use `std::thread` or `std::async` to start many enough thread

File: be/src/util/once.h (hunk 2 above)
Comment (yiguolei) [reply be/src/util/once.h:84]: No, if exception occured during fn, the flag should be false.

Comment (yiguolei) [reply be/src/util/once.h:101]: No, if stored then other thread will check the has called == true, and will access the result. It is wrong.

**Comment by yiguolei** [comment]
run buildall

**Review by Gabriel39 — APPROVED** [review]

**Review by mrhhsg — APPROVED** [review]
lgtm

**Pull request merged**